### PR TITLE
Remove user from keycloak server on BeforeUserRemovedEvent

### PR DESCRIPTION
**Pull request by vparfonov**
Signed-off-by: Vitalii Parfonov <vparfono@redhat.com>

### What does this PR do?
Remove user from Keycloak server on BeforeUserRemovedEvent 

### Screenshot/screencast of this PR

https://youtu.be/ZHlKZj8zXAw

### What issues does this PR fix or reference?

https://github.com/eclipse/che/issues/17541

### How to test this PR?

images for trying : 
- vparfonov/che-server:0915-3
- vparfonov/che-operator:0914

Steps: 
1. register new user in Che;
2. go to swagger page: https://{che-host}/swagger/#!/user;
3. get users;
4. find id of registered user;
5. delete user with id;
6. try to login to the Che
7. you will not be able todo that

### PR Checklist

[As the author of this Pull Request I made sure that:](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#pull-request-template-and-its-checklist)

- [x] [The Eclipse Contributor Agreement is valid](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#the-eclipse-contributor-agreement-is-valid)
- [x] [Code produced is complete](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#code-produced-is-complete)
- [x] [Code builds without errors](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#code-builds-without-errors)
- [ ] [Tests are covering the bugfix](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#tests-are-covering-the-bugfix)
- [ ] [The repository devfile is up to date and works](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#the-repository-devfile-is-up-to-date-and-works)
- [x] [Sections `What issues does this PR fix or reference` and `How to test this PR` completed](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#sections-what-issues-does-this-pr-fix-or-reference-and-how-to-test-this-pr-completed)
- [ ] [Relevant user documentation updated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#relevant-contributing-documentation-updated)
- [ ] [Relevant contributing documentation updated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#relevant-contributing-documentation-updated)
- [ ] [CI/CD changes implemented, documented and communicated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#cicd-changes-implemented-documented-and-communicated)

### Reviewers

Reviewers, please comment how you tested the PR when approving it.

### CODE REVIEW

**Comment by vparfonov** [comment]
[crw-ci-test]

File: assembly/assembly-wsmaster-war/src/main/webapp/WEB-INF/classes/che/multiuser.properties [hunk 1]
```diff
@@ -168,3 +168,7 @@ che.keycloak.username_claim=NULL
 # If set to "delegated", then the service will use Keycloak IdentityProvider mechanism.
 # Runtime Exception wii be thrown, in case if this property is not set properly.
 che.oauth.service_mode=delegated
+
+# Configuration for enabling removing user from Keycloak server on removing user from Che database.
+# By default should be false, will need only in some special cases
+che.keycloak.cascade_user_removal_enabled=true
```
Comment (monaka) [inline assembly/assembly-wsmaster-war/src/main/webapp/WEB-INF/classes/che/multiuser.properties]:
```suggestion
che.keycloak.cascade_user_removal_enabled=false
```

Comment (vparfonov) [reply assembly/assembly-wsmaster-war/src/main/webapp/WEB-INF/classes/che/multiuser.properties]: Thanks, for catching

File: assembly/assembly-wsmaster-war/src/main/java/org/eclipse/che/api/deploy/WsMasterModule.java [hunk 2]
```diff
@@ -373,6 +374,7 @@ private void configureMultiUserMode(
     install(new OrganizationJpaModule());
 
     install(new KeycloakModule());
+    install(new KeycloakUserRemoverModule()); // TODO remove before merge to the master
```
Comment (sleshchenko) [inline assembly/assembly-wsmaster-war/src/main/java/org/eclipse/che/api/deploy/WsMasterModule.java]: Please address that TODO

File: assembly/assembly-wsmaster-war/src/main/webapp/WEB-INF/classes/che/multiuser.properties (hunk 1 above)
Comment (sleshchenko) [reply assembly/assembly-wsmaster-war/src/main/webapp/WEB-INF/classes/che/multiuser.properties]: In which special cases it's going to be needed?

Comment (sleshchenko) [reply assembly/assembly-wsmaster-war/src/main/webapp/WEB-INF/classes/che/multiuser.properties]: I think it makes sense to add references to credentials here as well, like java class has:
```
For correct work need to set admin username ${che.keycloak.admin_username} and password
 * ${che.keycloak.admin_password}.
```

File: assembly/assembly-wsmaster-war/src/main/java/org/eclipse/che/api/deploy/WsMasterModule.java (hunk 2 above)
Comment (vparfonov) [reply assembly/assembly-wsmaster-war/src/main/java/org/eclipse/che/api/deploy/WsMasterModule.java]: _remove before merge to the master_

Comment (sleshchenko) [reply assembly/assembly-wsmaster-war/src/main/java/org/eclipse/che/api/deploy/WsMasterModule.java]: Why not remove it now? PR can't be ready to review if it has TODOs, can it?

File: multiuser/keycloak/che-multiuser-keycloak-user-remover/src/main/java/org/eclipse/che/multiuser/keycloak/server/KeycloakUserRemover.java [hunk 3]
```diff
@@ -0,0 +1,190 @@
...
+    String accessToken = "";
+    HttpURLConnection http = null;
+    try {
+      http = (HttpURLConnection) new URL(keycloakTokenEndpoint).openConnection();
+      http.setRequestMethod(POST);
+      http.setAllowUserInteraction(false);
+      http.setRequestProperty("Content-Type", "application/x-www-form-urlencoded");
+      http.setInstanceFollowRedirects(true);
+      http.setDoOutput(true);
+      OutputStream output = http.getOutputStream();
```
Comment (skabashnyuk) [inline multiuser/keycloak/che-multiuser-keycloak-user-remover/src/main/java/org/eclipse/che/multiuser/keycloak/server/KeycloakUserRemover.java]: Try-finally?

Comment (skabashnyuk) [reply multiuser/keycloak/che-multiuser-keycloak-user-remover/src/main/java/org/eclipse/che/multiuser/keycloak/server/KeycloakUserRemover.java]: connection timeout?

File: multiuser/keycloak/che-multiuser-keycloak-user-remover/src/main/java/org/eclipse/che/multiuser/keycloak/server/KeycloakUserRemover.java [hunk 4]
```diff
@@ -0,0 +1,194 @@
...
+          .append("&username=")
+          .append(keycloakUser)
+          .append("&password=")
+          .append(keycloakPassword)
+          .append("&client_id=admin-cli");
+      try (OutputStream output = http.getOutputStream()) {
+        output.write(sb.toString().getBytes(UTF_8));
+      }
+      if (http.getResponseCode() != 200) {
+        throw new RuntimeException(
```
[the remark this answers is not in the document]
Comment (sleshchenko) [reply multiuser/keycloak/che-multiuser-keycloak-user-remover/src/main/java/org/eclipse/che/multiuser/keycloak/server/KeycloakUserRemover.java]: I would like to see it addressed for RuntimeException below as well.

File: multiuser/keycloak/che-multiuser-keycloak-user-remover/src/main/java/org/eclipse/che/multiuser/keycloak/server/deploy/KeycloakUserRemoverModule.java [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+import com.google.inject.AbstractModule;
+import org.eclipse.che.inject.DynaModule;
+import org.eclipse.che.multiuser.keycloak.server.KeycloakUserRemover;
+import org.eclipse.che.multiuser.keycloak.server.KeycloakUserRemover.RemoveUserListener;
+
+@DynaModule
+public class KeycloakUserRemoverModule extends AbstractModule {
+  @Override
+  protected void configure() {
+    bind(KeycloakUserRemover.class);
```
Comment (sleshchenko) [inline multiuser/keycloak/che-multiuser-keycloak-user-remover/src/main/java/org/eclipse/che/multiuser/keycloak/server/deploy/KeycloakUserRemoverModule.java]: I believe it's redundant binding.

File: multiuser/keycloak/che-multiuser-keycloak-user-remover/src/main/java/org/eclipse/che/multiuser/keycloak/server/deploy/KeycloakUserRemoverModule.java [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+ *   Red Hat, Inc. - initial API and implementation
+ */
+package org.eclipse.che.multiuser.keycloak.server.deploy;
+
+import com.google.inject.AbstractModule;
+import org.eclipse.che.inject.DynaModule;
+import org.eclipse.che.multiuser.keycloak.server.KeycloakUserRemover;
+import org.eclipse.che.multiuser.keycloak.server.KeycloakUserRemover.RemoveUserListener;
+
+@DynaModule
```
Comment (sleshchenko) [inline multiuser/keycloak/che-multiuser-keycloak-user-remover/src/main/java/org/eclipse/che/multiuser/keycloak/server/deploy/KeycloakUserRemoverModule.java]: As far as I remember we tried to get rid of DynaModules to speed up container start up but I may be wrong.
So, I would see that module installed somewhere here https://github.com/eclipse/che/blob/9f4b49bdaa0eb58f8d9ee006e19b957d24b7c9d2/assembly/assembly-wsmaster-war/src/main/java/org/eclipse/che/api/deploy/WsMasterModule.java#L310

Comment (vparfonov) [reply multiuser/keycloak/che-multiuser-keycloak-user-remover/src/main/java/org/eclipse/che/multiuser/keycloak/server/deploy/KeycloakUserRemoverModule.java]: @skabashnyuk WDYT?

Comment (vparfonov) [reply multiuser/keycloak/che-multiuser-keycloak-user-remover/src/main/java/org/eclipse/che/multiuser/keycloak/server/deploy/KeycloakUserRemoverModule.java]: @sleshchenko Maybe need to deprecate `@DynaModule` if you decide to not use it anymore

File: multiuser/keycloak/che-multiuser-keycloak-user-remover/src/main/java/org/eclipse/che/multiuser/keycloak/server/KeycloakUserRemover.java [hunk 7]
```diff
@@ -0,0 +1,194 @@
...
+      @Nullable @Named("che.keycloak.admin_password") String keycloakPassword,
+      KeycloakSettings keycloakSettings,
+      HttpJsonRequestFactory requestFactory) {
+    this.keycloakUser = keycloakUser;
+    this.keycloakPassword = keycloakPassword;
+    this.requestFactory = requestFactory;
+    if (userRemovalEnabled) {
+      String serverUrl = keycloakSettings.get().get(AUTH_SERVER_URL_SETTING);
+      if (serverUrl == null) {
+        throw new RuntimeException(AUTH_SERVER_URL_SETTING + " is not configured");
```
[the remark this answers is not in the document]
Comment (vparfonov) [reply multiuser/keycloak/che-multiuser-keycloak-user-remover/src/main/java/org/eclipse/che/multiuser/keycloak/server/KeycloakUserRemover.java]: As you wish ...

File: multiuser/keycloak/che-multiuser-keycloak-user-remover/src/main/java/org/eclipse/che/multiuser/keycloak/server/KeycloakUserRemover.java [hunk 8]
```diff
@@ -0,0 +1,194 @@
...
+import org.eclipse.che.core.db.cascade.CascadeEventSubscriber;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Remove user from Keycloak server on {@link
+ * org.eclipse.che.api.user.server.event.BeforeUserRemovedEvent}. Turn on with
+ * ${che.keycloak.cascade_user_removal_enabled} property.
+ *
+ * <p>For correct work need to set admin username ${che.keycloak.admin_username} and password
```
[the remark this answers is not in the document]
Comment (vparfonov) [reply multiuser/keycloak/che-multiuser-keycloak-user-remover/src/main/java/org/eclipse/che/multiuser/keycloak/server/KeycloakUserRemover.java]: fixed

File: multiuser/keycloak/che-multiuser-keycloak-user-remover/src/main/java/org/eclipse/che/multiuser/keycloak/server/KeycloakUserRemover.java [hunk 9]
```diff
@@ -0,0 +1,194 @@
...
+import org.eclipse.che.commons.annotation.Nullable;
+import org.eclipse.che.commons.lang.IoUtil;
+import org.eclipse.che.core.db.cascade.CascadeEventSubscriber;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Remove user from Keycloak server on {@link
+ * org.eclipse.che.api.user.server.event.BeforeUserRemovedEvent}. Turn on with
+ * ${che.keycloak.cascade_user_removal_enabled} property.
```
[the remark this answers is not in the document]
Comment (vparfonov) [reply multiuser/keycloak/che-multiuser-keycloak-user-remover/src/main/java/org/eclipse/che/multiuser/keycloak/server/KeycloakUserRemover.java]: fixed

File: multiuser/keycloak/che-multiuser-keycloak-user-remover/src/main/java/org/eclipse/che/multiuser/keycloak/server/KeycloakUserRemover.java [hunk 10]
```diff
@@ -0,0 +1,194 @@
...
+      }
+      if (Strings.isNullOrEmpty(keycloakUser) || Strings.isNullOrEmpty(keycloakPassword)) {
+        throw new RuntimeException("Keycloak administrator username or password not set.");
+      }
+      this.keycloakTokenEndpoint = serverUrl + "/realms/master/protocol/openid-connect/token";
+      this.keycloakRemoveUserUrl = serverUrl + "/admin/realms/" + realm + "/users/";
+    }
+  }
+
+  /** */
```
[the remark this answers is not in the document]
Comment (vparfonov) [reply multiuser/keycloak/che-multiuser-keycloak-user-remover/src/main/java/org/eclipse/che/multiuser/keycloak/server/KeycloakUserRemover.java]: added

File: multiuser/keycloak/che-multiuser-keycloak-user-remover/src/main/java/org/eclipse/che/multiuser/keycloak/server/KeycloakUserRemover.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (vparfonov) [reply multiuser/keycloak/che-multiuser-keycloak-user-remover/src/main/java/org/eclipse/che/multiuser/keycloak/server/KeycloakUserRemover.java]: done

File: multiuser/keycloak/che-multiuser-keycloak-user-remover/src/main/java/org/eclipse/che/multiuser/keycloak/server/KeycloakUserRemover.java [hunk 11]
```diff
@@ -0,0 +1,200 @@
...
+import java.io.IOException;
+import java.io.InputStreamReader;
+import java.io.OutputStream;
+import java.net.HttpURLConnection;
+import java.net.URL;
+import javax.annotation.PostConstruct;
+import javax.annotation.PreDestroy;
+import javax.inject.Inject;
+import javax.inject.Named;
+import javax.inject.Singleton;
```
Comment (mshaposhnik) [inline multiuser/keycloak/che-multiuser-keycloak-user-remover/src/main/java/org/eclipse/che/multiuser/keycloak/server/KeycloakUserRemover.java:34]:
```suggestion
import javax.inject.Singleton;
import javax.ws.rs.core.HttpHeaders;
import javax.ws.rs.core.MediaType;
```

File: multiuser/keycloak/che-multiuser-keycloak-user-remover/src/main/java/org/eclipse/che/multiuser/keycloak/server/KeycloakUserRemover.java [hunk 12]
```diff
@@ -0,0 +1,200 @@
...
+  private String requestToken() throws ServerException {
+    String accessToken;
+    HttpURLConnection http = null;
+    try {
+      http = (HttpURLConnection) new URL(keycloakTokenEndpoint).openConnection();
+      http.setConnectTimeout(60000);
+      http.setReadTimeout(60000);
+      http.setRequestMethod(POST);
+      http.setAllowUserInteraction(false);
+      http.setRequestProperty("Content-Type", "application/x-www-form-urlencoded");
```
Comment (mshaposhnik) [inline multiuser/keycloak/che-multiuser-keycloak-user-remover/src/main/java/org/eclipse/che/multiuser/keycloak/server/KeycloakUserRemover.java]:
```suggestion
      http.setRequestProperty(HttpHeaders.CONTENT_TYPE, MediaType.APPLICATION_FORM_URLENCODED);
```

Comment (mshaposhnik) [reply multiuser/keycloak/che-multiuser-keycloak-user-remover/src/main/java/org/eclipse/che/multiuser/keycloak/server/KeycloakUserRemover.java]: but in general i would better use `doRequest` from `KeycloakServiceClient` (with minor mod-s)

**Review by mshaposhnik — APPROVED** [review]

File: multiuser/keycloak/che-multiuser-keycloak-user-remover/src/main/java/org/eclipse/che/multiuser/keycloak/server/KeycloakUserRemover.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (sleshchenko) [reply multiuser/keycloak/che-multiuser-keycloak-user-remover/src/main/java/org/eclipse/che/multiuser/keycloak/server/KeycloakUserRemover.java]: It still seems to be RuntimeException

File: multiuser/keycloak/che-multiuser-keycloak-user-remover/src/main/java/org/eclipse/che/multiuser/keycloak/server/KeycloakUserRemover.java [hunk 13]
```diff
@@ -0,0 +1,202 @@
...
+          .append(keycloakUser)
+          .append("&password=")
+          .append(keycloakPassword)
+          .append("&client_id=admin-cli");
+      try (OutputStream output = http.getOutputStream()) {
+        output.write(sb.toString().getBytes(UTF_8));
+      }
+      if (http.getResponseCode() != 200) {
+        throw new RuntimeException(
+            "Cannot get access KeyCloak token. Server response code: "
```
Comment (sleshchenko) [inline multiuser/keycloak/che-multiuser-keycloak-user-remover/src/main/java/org/eclipse/che/multiuser/keycloak/server/KeycloakUserRemover.java]:
```suggestion
            "Cannot get Keycloak access token. Server response: "
```

File: multiuser/keycloak/che-multiuser-keycloak-user-remover/src/main/java/org/eclipse/che/multiuser/keycloak/server/KeycloakUserRemover.java [hunk 14]
```diff
@@ -0,0 +1,202 @@
...
+          new BufferedReader(new InputStreamReader(http.getInputStream(), UTF_8));
+
+      JsonParser jsonParser = new JsonParser();
+      JsonElement jsonElement = jsonParser.parse(response);
+      JsonObject asJsonObject = jsonElement.getAsJsonObject();
+      if (asJsonObject.has("access_token")) {
+        accessToken = asJsonObject.get("access_token").getAsString();
+      } else {
+        LOG.error("Cannot get access KeyCloak token from server response");
+        throw new ServerException("Cannot get access KeyCloak token from server response");
```
Comment (sleshchenko) [inline multiuser/keycloak/che-multiuser-keycloak-user-remover/src/main/java/org/eclipse/che/multiuser/keycloak/server/KeycloakUserRemover.java]:
```suggestion
        throw new ServerException("Cannot get Keycloak access token from server response");
```

File: multiuser/keycloak/che-multiuser-keycloak-user-remover/src/main/java/org/eclipse/che/multiuser/keycloak/server/KeycloakUserRemover.java [hunk 15]
```diff
@@ -0,0 +1,202 @@
...
+      }
+    } catch (IOException | JsonSyntaxException ex) {
+      LOG.error(ex.getMessage(), ex);
+      throw new ServerException("Cannot get access KeyCloak token.", ex);
```
Comment (sleshchenko) [inline multiuser/keycloak/che-multiuser-keycloak-user-remover/src/main/java/org/eclipse/che/multiuser/keycloak/server/KeycloakUserRemover.java]:
```suggestion
      throw new ServerException("Cannot get Keycloak access token.", ex);
```

File: multiuser/keycloak/che-multiuser-keycloak-user-remover/src/main/java/org/eclipse/che/multiuser/keycloak/server/KeycloakUserRemover.java [hunk 16]
```diff
@@ -0,0 +1,202 @@
...
+      final BufferedReader response =
+          new BufferedReader(new InputStreamReader(http.getInputStream(), UTF_8));
+
+      JsonParser jsonParser = new JsonParser();
+      JsonElement jsonElement = jsonParser.parse(response);
+      JsonObject asJsonObject = jsonElement.getAsJsonObject();
+      if (asJsonObject.has("access_token")) {
+        accessToken = asJsonObject.get("access_token").getAsString();
+      } else {
+        LOG.error("Cannot get access KeyCloak token from server response");
```
Comment (sleshchenko) [inline multiuser/keycloak/che-multiuser-keycloak-user-remover/src/main/java/org/eclipse/che/multiuser/keycloak/server/KeycloakUserRemover.java]: Why you need that? In any failure, you're supposed to log error here https://github.com/eclipse/che/pull/17858/files#diff-6faa75cae54e91254f0b4f0d3d8b9b8bR115

File: multiuser/keycloak/che-multiuser-keycloak-user-remover/src/main/java/org/eclipse/che/multiuser/keycloak/server/KeycloakUserRemover.java [hunk 17]
```diff
@@ -0,0 +1,202 @@
...
+        throw new ServerException("Cannot get access KeyCloak token from server response");
+      }
+    } catch (IOException | JsonSyntaxException ex) {
+      LOG.error(ex.getMessage(), ex);
```
Comment (sleshchenko) [inline multiuser/keycloak/che-multiuser-keycloak-user-remover/src/main/java/org/eclipse/che/multiuser/keycloak/server/KeycloakUserRemover.java:160]: I think it's redundant

File: multiuser/keycloak/che-multiuser-keycloak-user-remover/src/main/java/org/eclipse/che/multiuser/keycloak/server/KeycloakUserRemover.java [hunk 18]
```diff
@@ -0,0 +1,202 @@
...
+      String token = requestToken();
+      int responseCode =
+          requestFactory
+              .fromUrl(keycloakRemoveUserUrl + userId)
+              .setAuthorizationHeader("Bearer " + token)
+              .useDeleteMethod()
+              .request()
+              .getResponseCode();
+      if (responseCode != 204) {
+        throw new ServerException("Can't remove user from KeyClock. UserId:" + userId);
```
Comment (sleshchenko) [inline multiuser/keycloak/che-multiuser-keycloak-user-remover/src/main/java/org/eclipse/che/multiuser/keycloak/server/KeycloakUserRemover.java]:
```suggestion
        throw new ServerException("Can't remove user from Keycloak. UserId:" + userId);
```

**Review by sleshchenko — APPROVED** [review]
It does not seem to be related to your PR but still build job fails with:
> [ERROR] Tests run: 11, Failures: 1, Errors: 0, Skipped: 0, Time elapsed: 1.931 s <<< FAILURE! - in TestSuite
> [ERROR] > shouldRecordCronExecutorServiceMetrics(org.eclipse.che.commons.observability.MeteredExecutorServiceWrapperTest)  Time elapsed: 0.279 s  <<< FAILURE!
> java.lang.AssertionError: did not expect to find [1.0] but found [0.0]
>	at org.eclipse.che.commons.observability.MeteredExecutorServiceWrapperTest.shouldRecordCronExecutorServiceMetrics(MeteredExecutorServiceWrapperTest.java:358)

File: multiuser/keycloak/che-multiuser-keycloak-user-remover/src/main/java/org/eclipse/che/multiuser/keycloak/server/KeycloakUserRemover.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (vparfonov) [reply multiuser/keycloak/che-multiuser-keycloak-user-remover/src/main/java/org/eclipse/che/multiuser/keycloak/server/KeycloakUserRemover.java]: done

**Comment by vparfonov** [comment]
[crw-ci-test]

**Comment by vparfonov** [comment]
[crw-ci-test]

**Comment by vparfonov** [comment]
[crw-ci-test]

**Comment by dmytro-ndp** [comment]
[crw-ci-test]

File: multiuser/keycloak/che-multiuser-keycloak-user-remover/src/main/java/org/eclipse/che/multiuser/keycloak/server/deploy/KeycloakUserRemoverModule.java (hunk 6 above)
Comment (skabashnyuk) [reply multiuser/keycloak/che-multiuser-keycloak-user-remover/src/main/java/org/eclipse/che/multiuser/keycloak/server/deploy/KeycloakUserRemoverModule.java]: @sleshchenko probably right. Let's replace @DynaModule with an explicit entry in `WsMasterModule.java`

Comment (vparfonov) [reply multiuser/keycloak/che-multiuser-keycloak-user-remover/src/main/java/org/eclipse/che/multiuser/keycloak/server/deploy/KeycloakUserRemoverModule.java]: ok

File: assembly/assembly-wsmaster-war/src/main/webapp/WEB-INF/classes/che/multiuser.properties [hunk 19]
```diff
@@ -168,3 +168,16 @@ che.keycloak.username_claim=NULL
...
 che.oauth.service_mode=delegated
+
+# Configuration for enabling removing user from Keycloak server on removing user from Che database.
+# By default it's disabled. Can be enabled in some special cases when deleting a user in Che database
+# should execute removing related-user from Keycloak.
+# For correct work need to set admin username ${che.keycloak.admin_username} and password
+# ${che.keycloak.admin_password}.
+che.keycloak.cascade_user_removal_enabled=false
+
+# Keycloak admin username
```
[the remark this answers is not in the document]
Comment (vparfonov) [reply assembly/assembly-wsmaster-war/src/main/webapp/WEB-INF/classes/che/multiuser.properties]: done

File: assembly/assembly-wsmaster-war/src/main/webapp/WEB-INF/classes/che/multiuser.properties [hunk 20]
```diff
@@ -168,3 +168,16 @@ che.keycloak.username_claim=NULL
...
+che.keycloak.admin_username=NULL
+
+# Keycloak admin password
```
[the remark this answers is not in the document]
Comment (vparfonov) [reply assembly/assembly-wsmaster-war/src/main/webapp/WEB-INF/classes/che/multiuser.properties]: done

**Review by skabashnyuk — APPROVED** [review]

**Pull request merged**